### ISSUE TITLE
Support optional compression of remote state JSON

### ISSUE THREAD

**Issue by costela** [pulumi/pulumi#5469]
When using a remote state backend, some resources might cause the state file to grown significantly, causing `pulumi up` to take 10s of minutes instead of seconds.

One such example is using `yaml.ConfigFile` to deploy [cert-manager](/jetstack/cert-manager)'s single YAML deployment to kubernetes.
The YAML file defines a series of kubernetes CRDs with accompanying documentation, which means it weights a hefty 1.6Mb. Applying it leads to a 10Mb JSON state file (from ~500kb), which gets read and written for each new resource defined in the file, causing the whole operation to take a significant amount of time. 

The remote state file is currently stored as plain JSON. Compressing it with gzip would already speed things up significantly.

**Comment by lukehoban** [comment]
Which backend are you using?  The default Pulumi Service backend does compress the state file both in transit and at rest. Are you using the local filesystem or S3/GCS/blob storage?

**Comment by costela** [comment]
ah yes, forgot to mention: I'm using GCS.

I understand using Pulumi Service is probably preferable, from a "Pulumi Corp" perspective, but it would still be nice if other backends also got a bit of love :wink:

**Comment by lukehoban** [comment]
> I understand using Pulumi Service is probably preferable, from a "Pulumi Corp" perspective, but it would still be nice if other backends also got a bit of love 😉

Definitely!  Just wanted to make sure as I suspect the answer is specific to each backend and how the cloud provider APIs manage the compression in transit.

> One such example is using yaml.ConfigFile to deploy cert-manager's single YAML deployment to kubernetes.
The YAML file defines a series of kubernetes CRDs with accompanying documentation, which means it weights a hefty 1.6Mb. Applying it leads to a 10Mb JSON state file (from ~500kb), which gets read and written for each new resource defined in the file, causing the whole operation to take a significant amount of time.

Note that this explosion in size for something that is just one line of code in Pulumi likely will warrant even deeper improvements to help tame this.  Compression might help - but I suspect there will be issues related to this even if things are being compressed in transit.  Not yet sure exactly what else we will need to do - but worth calling out this potentially broader need as well.

**Comment by costela** [comment]
> Note that this explosion in size for something that is just one line of code in Pulumi likely will warrant even deeper improvements to help tame this. Compression might help - but I suspect there will be issues related to this even if things are being compressed in transit. Not yet sure exactly what else we will need to do - but worth calling out this potentially broader need as well.

oh yes, makes sense. Compression just seemed like a relatively low-hanging fruit to at least mitigate the most obvious time-sink.

**Comment by awoimbee** [comment]
Some of my stacks json files a >16MiB.
Pulumi is super slow, one needs a fast connection to dare touch these stacks. I use S3, my bucket went above 150GB before I removed the stack history. :/
Using the gzip command, a >16MiB stack becomes <1MiB, gzip should have been used since the beginning.

I looked at the filestate backend, it's a bit spaghetti :/.
Still, if a PR were to land with gzip compression, should there be a toggle to disable it ?

**Comment by jsravn** [comment]
Alternatively we could just compress the transport, right? That should be a really minimal change with a huge benefit for us masochistic self hosted users.

**Comment by jsravn** [comment]
> Alternatively we could just compress the transport, right? That should be a really minimal change with a huge benefit for us masochistic self hosted users.

Just realised this is up to the capabilities of the cloud provider, in our case GCS doesn't support compression in flight. So not much Pulumi can do about it. The next best thing would be compressing the backend, at least until self hosted states are reworked not to upload on every change.
